File: tests/station-test-light/hw/analog_inputs.py
# analog_inputs.py - A class for reading the values of the analog lines
#
# AnalogInputs provides a set of methods for reading the voltages on the 
# analog input lines connected to the ADC. 
#
# There are 8 analog lines connected to the ADC. The first four (lines 0-3)
# are connected to the center tap on a 10kOhm potentiometer wired between 
# ground and Vcc (+5V). They will read values between 0 and 5V, depending on
# the position of the knob. The second four (lines 4-7) are used as jumpers,
# and will read either around 0 volts, or around 5 volts (things aren't 
# exact here).
#
# This class is a simple wrapper around the ads1256-py library, so see that
# for details on how to use the ADC.
#
# Depends on the ads1256 Python module and the supporting C library.

import ads1256

class AnalogInputs:
    # A state variable to ensure that we only initialize the ADS1256 once.
    initialized = False

    # Constructor for an instance of the class. Takes no parameters.
    def __init__(self):
        # If the ADS1256 hasn't been initialized yet, initialize it and
        # configure it.
        if AnalogInputs.initialized == False:
            ads1256.initialize()

            # We configure it with a gain of 1 (no gain), and to sample at
            # 1000 samples per second (SPS).
            ads1256.configure(ads1256.ADS1256_GAIN_1, ads1256.ADS1256_1000SPS)

            # Set the flag to indicate that we've been initialized.
            AnalogInputs.initialized = True

    # Read the current voltage values from the ADC. 
    #
    # Returns a list of 8 values.
    def read_values(self):
        # We're doing single-ended (not differential) analog conversions,
        # and we want to read channels from 0 to the maximum.
        return ads1256.read_channels(0, 
                                     ads1256.ADS1256_SINGLE_CHANNEL_COUNT - 1, 
                                     ads1256.ADS1256_INPUT_SINGLE_ENDED)

File: tests/old/station-test-ruby/lib/analog_inputs.rb
# analog_inputs.rb - A class for reading the values of the analog lines
#
# $Id: $

# AnalogInputs provides a set of methods for reading the voltages on the 
# analog input lines connected to the ADC. 
#
# There are four analog lines connected to the ADC, each connected to the
# center tap on a 10kOhm potentiometer wired between ground and Vcc (+5V).
#
# The ADC is connected to the Raspberry Pi via I2C. This class depends on 
# a utility program to read the values. That program returns a comma-separated
# list of values between -32,768 and +32,767, representing voltages between
# -6.144 and +6.144 volts.
#
# The program returns the values for all 4 input lines at once. Those values
# are cached, and will be returned until the values are re-read by calling
# read_values().
#
# == Example
#
#   require 'analog_inputs'
#
#   analog_inputs = AnalogInputs.new
#
#   voltage = analog_inputs.get_value(1)
#   puts "Analog line 1 is at #{voltage*6.144/32767} volts."
#
#   sleep 1
#
#   analog_inputs.read_values
#   voltage = analog_inputs.get_value(1)
#   puts "Analog line 1 is now at #{voltage*6.144/32767} volts."

class AnalogInputs
  # The location of the ADC helper tool.
  @@ADC_INPUT_PROGRAM = "/home/pi/harmoniscope/a2d-test/read_all -r"

  # The most recent set of values read from the ADC
  @cur_values

  # Initialize a new instance. Currently does nothing.
  def initialize()
  end
 
  # Read the current voltage values from the ADC. 
  def read_values()
    output = `#{@@ADC_INPUT_PROGRAM}`
    raise "Error reading ADC values (#{$?.exitstatus})" if ($?.exitstatus != 0)

    @cur_values = output.split(',').map(&:to_i)

    unless @cur_values.length == 4
      raise "Incorrect number of analog inputs (#{@cur_values.length})" 
    end
  end

  # Get the voltage value that was read in for the given line_number.
  def get_value(line_number)
    line_number = line_number.to_i

    unless line_number >= 0 && line_number < 4
      raise "Invalid analog line number #{line_number}" 
    end

    return @cur_values[line_number]
  end
end

